lesson-2/function-return-function.md

```diff
@@ -0,0 +1,129 @@
+    removeDouble(1)()
+  }
+}
+```
+
+What happened here? Having the nested function actually made things worse! But wait, there is more. If we wish to pass the function as an event handler, things turn the other way around:
+
+```javascript
+render() {
+  const remover = this.removeSingle.bind(this, noteId)
+  return (
+    <div>
+      {/* anonymous function; fat arrow binds `this` */}
+      <button onClick={(e) => { this.removeSingle(noteId) }}>
+
+      {/* ... or bound function */}
+      <button onClick={remover}>
+
+      {/* but! */}
+      <button onClick={this.removeDouble(noteId)}>
+    </div>
+  )
+}
+```
+
+See? Function that returns function made it easier to use in event handlers. Which happens a lot in frontend code; you only want to hook event handlers (some call them event listeners) and wait for user action.
+
+Some more reading:
+
+- [Eloquent JavaScript: Higher-Order Functions](https://eloquentjavascript.net/05_higher_order.html)
+- [Davis Walsh: JavaScript Functions that Return Functions](https://davidwalsh.name/javascript-functions)
+
+This concept is somehow but not exactly related to currying (homework: _why?_). Now that you understand the concept of functions returning functions you may check it out too:
+
+- [M. David Green: Currying in Functional JavaScript](https://www.sitepoint.com/currying-in-functional-javascript/)
+- [Alex Rauschmayer (2ality): Currying versus partial application](http://2ality.com/2011/09/currying-vs-part-eval.html)
+
+## React, classes and misbehaving `this`
+
+This is not directly related to the original question but I want to include it here because is important to understand the most common problem with classes in JS: `this`.
+
+When using classes you need to be careful what exactly is `this` in your code. It may or may not be what you expect. See this code:
+
+```javascript
+class ThisIsHorrible {
+
+  // defined as a oldschool "method"
+  brokenFn() { 
+    console.log(this, this.javascript)
+    // throws a TypeError: this.setState is not a function
+    this.setState({ output: this.javascript })
+  }
+
+  // defined with =
+  workingFn = () => {
+    console.log(this, this.javascript)
+    // works as expected
+    this.setState({ output: this.javascript })
+  }
+
+}
+```
+
+(I have omitted some code for clarity. See the full running [example on codepen.io](https://codepen.io/corkscreewe/pen/vpBwaM?editors=0010) if you're interested.)
+
+For complete mastery and understanding of various rules for `this` I recommend the excellent book [You Don't Know JS: this & Object Prototypes](https://github.com/getify/You-Dont-Know-JS/blob/master/this%20%26%20object%20prototypes/README.md).
+
+>Side-Side note: Why did I put "method" in quotes? Because these are not methods. They may look like one but are functions. `{ foo() {} }` [is a shorthand](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Method_definitions) for `{ foo: function foo() {} }`.
```

Comment:
well, a function which is a property of an object and is bound to that object is a method, so "methods" are real OOP methods - do you have a definition by which "these are not methods" please?

Reply:
The way I understand it a "method" is always bound to an instance of its class. JavaScript does not have classes (regardless of the `class` keyword) and therefore has no methods. These are functions that happen to be attached to this object, but are in no way bound to it:

```javascript
class ThisIsNotAClass {
  constructor(name) {
    this.name = name
  }
  
  notAMethod() {
    console.log("Hello, my name is " + this.name)
  }
}

const instance = new ThisIsNotAClass("Method")
instance.notAMethod()

const person = {
  name: "Peter",
  greet: instance.notAMethod
}
person.greet()
```

Reply:
of course JS does have methods - inherited from the prototype and bound to the object itself, e.g. `"abc".toUpperCase()` - see the wording in [the docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/toUpperCase)

`new` creates a new object and dynamically binds that new object to identifier `this` which becomes a special property of the (other) object returned by the constructor

while `"abc"` string literal is not created with the `new` keyword, it works in a similar way, the string object has methods, not "just" functions (of course methods are functions, but functions themselves are objects with methods such as `call` implementing the `()` syntax, not sure if explaining methods away as functions is really simplifying the stuff)

there is the difference between classical and prototypical inheritance, yes...

also assignment passes the reference to the function-object-which-is-a-method-of-another-object  without keeping the bond to that another-object:

`var unboundToUpperCase = "abc".toUpperCase`

and passing arguments in a function call is just assignment, and using JSX properties is just passing arguments in a function call after JSX is compiled to JS

Reply:
maybe in other words:

method = a function that is a property of an object

when a function is assigned to a variable (or used as a function argument or in JSX attribute), it is just a function, no longer the property of the previous object

?

Reply:
actually, this answer is better: https://stackoverflow.com/a/20279485/1176601 :)

Reply:
>method = a function that is a property of an object

Following this definition JS functions are methods too. But that is not how I would define it.

I'll probably remove the side-side note since it can be more confusing than useful.